Probing for more reliable route fee estimation

hieblmi opened the pull request:
This PR implements https://github.com/lightningnetwork/lnd/issues/7916

### Overview

In this PR we extend the implementation of the graph based`EstimateRouteFee` RPC with a payment probe based approach to obtain a more reliable routing fee estimate.

### Notes on implementation

- The new fee estimation API now optionally accepts an invoice from the target node and a probe payment timeout.
- Probing distinguishes two cases, (1) probing to a public destination key specified in the invoice or (2), probing a private invoice destination reachable through public hop hints in the invoice.
- If multiple hop hints are provided then, for each of the public ones, a probe is dispatched and each of the fee estimates is returned. Private hints are skipped.
- `lncli` adds support to use the graph-based and probe-based fee estimation:

graph-based: 
```
lncli estimateroutefee --dest 0266a18ed969ef95c8a5aa314b443b2b3b8d91ed1d9f8e95476f5f4647efdec079 --amt 100000
```
probe-based: 
```
lncli estimateroutefee reg_zane estimateroutefee --pay_req lnbcrt1m1pj... --timeout 60s
```

JssDWt commented:
There is currently a problem with prepayment probing in combination with some lightning service providers. Basically prepayment probes are not compatible with the current LSP protocols. Since nodes behind a LSP can be a significant portion of destinations, I think it's good to keep the following considerations in mind. In below context, assume the payee is a lightning node behind a LSP.

In the current state I generally recommend against sending prepayment probes when a LSP is involved. I guess the question is: how do you determine whether a LSP is involved? Let's outline the protocols and their caveats.

#### Present: Breez LSP
- Prepayment probes with a random payment hash will, in some situations around just-in-time channels, fail with `unknown_next_peer` returned by the LSP, even though the _actual_ payment with the real payment hash would succeed. 
- There is an option to re-hash the payment hash prefixed with `probing-01`, like [here](https://github.com/breez/lspd/blob/33b7e1dd6d53e26dd8bf1d9c08cd69bf085d4bd4/itest/probing_test.go#L54-L57). When used with the current Breez LSP, the Breez LSP will return `incorrect_or_unknown_payment_details` in the same situations as where the previous bullet would return `unknown_next_peer`.

#### Soon: LSPS2
Very soon Breez and multiple other LSPs will implement the [LSPS2 protocol](https://github.com/BitcoinAndLightningLayerSpecs/lsp/blob/main/LSPS2/README.md). This is a protocol for just-in-time channels. This initial version of the protocol is incompatible with prepayment probes. There are no payment hash tricks here.

These nodes should have feature bit 729 (`option_supports_lsps`) set in their node announcement. Recommendation is to send the prepayment probe to the node in the route hint, rather than the destination node, if this feature bit is set for the route hint node and the destination node is only reachable through the route hint node (quite a mouthful, sorry about that).

#### Medium term: LSPS4
The LSP spec group is currently in the process of designing a LSP protocol LSPS4 for just-in-time channels. When LSPS4 is fully implemented, prepayment probes won't be much of an issue. This is not going to be implemented soon, however.

#### Other thoughts
Prepayment probes don't work together with MPP very well. At least there is no indication whether you were able to deliver the full amount for the specified fee, even if all payment parts return with `incorrect_or_unknown_payment_details`. Point is the parts do not all arrive at the same time, like they would in an actual payment. The previous parts locking up liquidity in the route on an actual payment may very well change the path chosen for next parts.

alexbosworth commented on lnrpc/routerrpc/router.proto:
```diff
@@ -325,14 +325,28 @@ message TrackPaymentsRequest {
+    An invoice of the target node that the route fee request is intended for.
+    Its parameters are input to probe payments that estimate routing fees.
+     */
+    string invoice = 3;
```
normally these are called `payment_request` when it's a bolt 11 payreq input

alexbosworth commented on lnrpc/routerrpc/router.proto:
```diff
@@ -348,6 +362,11 @@ message RouteFeeResponse {
+    /*
+    An indication whether a probing payment reached the target node.
+     */
+    bool dest_reachable = 3;
```
shouldn't the fee be returned here?

hieblmi replied:
yes, the `routing_fee_msat` was already part of this message, see https://github.com/lightningnetwork/lnd/pull/8136/files/a0e152261696bef4197f7aaa7ed9068295b334f4#diff-35316b62bc12f39d4bf2426add7562a564d7deee49b9f9644f42766abb672c80R357

hieblmi commented:
Thanks for the detailed rundown on LSP mechanics @JssDWt.

The primary focus of this PR is to improve the currently active graph based method of estimating routing fees. It is unreliable because the gossip announcements might not accurately reflect the network topology or fees at the time when a payment occurs. So in this first step the desired outcome is to improve the fee estimation for payments to publicly known nodes.

Follow-up PRs will focus on MPP/AMP as noted in the linked issue https://github.com/lightningnetwork/lnd/issues/7916. This seems to have gotten possible since the release of AMP by invalidating the last shard of a AMP, see https://github.com/lightningnetwork/lnd/issues/4219#issuecomment-908833773.

I think LSP specific probing protocols will also land with a separate PR.

bitromortac commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -373,52 +380,120 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+	errChan := make(chan error, 1)
+	defer close(errChan)
+	go func() {
+		err := s.SendPaymentV2(req, nil)
```
I'm not sure if passing nil in here is the best solution. There is only a single `probeResultChan`, it gets reused for different `EstimateRouteFee` calls, so messages could interfere?

Perhaps passing in a stream implementation and then consume from it could work. I quickly tried it and the itest passed. Not sure if more of the `Router_SendPaymentV2Server` needs to be implemented. That way also the context could be reused.

```go
// probePaymentStream is a custom implementation of the grpc.ServerStream
// interface. It is used to send payment status updates to the caller on the
// stream channel.
type probePaymentStream struct {
	Router_SendPaymentV2Server

	stream chan *lnrpc.Payment
	ctx    context.Context
}

// Implement the grpc.ServerStream interface for probePaymentStream.
func (p *probePaymentStream) Send(response *lnrpc.Payment) error {
	p.stream <- response

	return nil
}

// Context returns the context of the stream.
func (p *probePaymentStream) Context() context.Context {
	return p.ctx
}

```

bitromortac reviewed (commented):
Nice, I think the api looks good, except that we could return an `lnrpc.PaymentFailureReason` to find out more about why a probe failed. Also added a suggestion for how to handle the streaming responses differently.

Another important point is backwards compatibility, I think we need to support the old ux for some time, as the new api now has a time component. So I'd suggest to keep the old code with the pubkey/amount input and only use the new estimation if an invoice is provided. This can then be removed after some time when the rpc fields are removed.

Roasbeef commented:
> Prepayment probes don't work together with MPP very well. At least there is no indication whether you were able to deliver the full amount for the specified fee, 

AMP solves that. You can send the N-1 shards, then make the Nth shared fail by giving it an invalid secret share.

Roasbeef commented:
@JssDWt 

> These nodes should have feature bit 729 (option_supports_lsps) set in their node announcement. Recommendation is to send the prepayment probe to the node in the route hint, rather than the destination node, if this feature bit is set for the route hint node and the destination node is only reachable through the route hint node (quite a mouthful, sorry about that).

How will that work if a probing hash is used, and also only a single shard is sent that may be below the actual payment amount? The LSP just decides on a value to use for channel opening? What if the sender doesn't retry again, the LSP just eats the chain fees cost? What about if the user already has a channel open?

JssDWt commented:
@Roasbeef 

Assuming you're asking about probing all the way to the destination node, and not up until the LSP. Let me answer these one by one.
 
> How will that work if a probing hash is used, and also only a single shard is sent that may be below the actual payment amount? 

In that case the single shard is failed back by the LSP after a timeperiod similar to the MPP timeout. The recipient's invoice amount is the threshold for letting htlcs through and trigger channel open.

> The LSP just decides on a value to use for channel opening? 

The recipient communicates the intended amount to receive to the LSP before creating the invoice. The invoice then contains a route hint with a scid specifically for this payment/channel order. 

> What about if the user already has a channel open?

In that case there are no issues. The destination works like any other lightning node. Only if the recipient ordered a cchannel for that payment will this LSPS2 spec be used.

Roasbeef commented:
@hieblmi is this ready to leave draft?

hieblmi commented:
> @hieblmi is this ready to leave draft?

@Roasbeef I will move to 'ready' once I have incorporated the hop hint fee estimation and refactor suggestions. Will focus on it.

lightninglabs-deploy commented:
@bitromortac: review reminder
@hieblmi, remember to re-request review from reviewers when ready

bitromortac commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -404,8 +437,189 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+
+	hopFee := hopHint.HopFee(int64(*payReq.MilliSat))
+
+	probeRequest.AmtMsat += hopFee
```
nice :+1:

bitromortac reviewed (commented):
Nice work @hieblmi :fire:. I left a comment to suggest how we could handle a situation with multiple route hints, as we probably need to distinguish route hints with different pubkeys from the case where only a single pubkey is present (possible LSP).

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -404,8 +437,189 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			},
+		}}
+	}
+	hopHint := payReq.RouteHints[0][0]
```
Right, thus far I only considered a single route hint.

I will add the ability to probe multiple route hints. In that case, would it make sense to return an array of estimates where each estimate, besides the estimated fee, would also have the public hop within a hint set?

hieblmi commented on lnrpc/routerrpc/router.proto:
```diff
@@ -343,14 +343,26 @@ message TrackPaymentsRequest {
+    A user preference of how long a probe payment should maximally be allowed to
+    take, denoted in seconds.
+     */
+    int32 timeout = 4;
```
The `SendPaymentRequest` tracks the timeout in seconds, hence I chose the same unit here.

ProofOfKeags replied:
Yeah and I also doubt we'll ever want a timeout that is shorter than a second considering global E2E round trip single TCP communication is 150ms. This is with no packet loss or without considering situations where we need to have multiple round trips where we try multiple routes. And since this latency is primarily bound by the speed of light, I don't anticipate it changing 😁

ProofOfKeags replied:
+1 for uint32

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -404,8 +437,189 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	// invoice destination as a route hint for convenience to consume it for
+	// probing further down.
+	if len(payReq.RouteHints) == 0 {
+		payReq.RouteHints = [][]zpay32.HopHint{{
```
I refactored the code, this section is gone.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -356,20 +358,51 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+		return nil, errors.New("cannot specify both a destination " +
+			"and an invoice")
+
+	case isProbeDestination && len(req.Dest) != 33:
```
Isn't it better to signal to the user which parameters weren't provided correctly?

ProofOfKeags replied:
this stuff would all be simplified by validating that one of these parameters was specified first. Then we can make assumptions about if it's specified at all then we validate it. From there you can have a single logic tree for validating destinations and invoices and keep the control flow fairly linear.

hieblmi replied:
I wanted to keep the vertical distance in the cases small so I needed more cases. But can just have cases for destination and invoices and then do the validation there.

hieblmi commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,75 @@
+
+	// Make sure Alice has heard it.
+	ht.AssertChannelPolicyUpdate(
+		mts.alice, mts.dave, expectedPolicy, chanPointAliceDave, false,
```
I think this is still correct as we try to ensure that Alice sees the high fee adjustment of Dave, so that we can be confident that our probe runs through Alice-Carol-Bob.

hieblmi commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,75 @@
+		lnrpc.PaymentFailureReason_FAILURE_REASON_INSUFFICIENT_BALANCE,
+		resp.FailureReason,
+	)
+
```
I've added two route hint tests and might add a few more edge cases. But for a review of the test frame work the next set of changes should be sufficient.

hieblmi commented on docs/release-notes/release-notes-0.18.0.md:
```diff
@@ -155,9 +155,16 @@
   [MinConf](https://github.com/lightningnetwork/lnd/pull/8097)(minimum number
   of confirmations) has been added to the `WalletBalance` RPC call.
+ 
+* [EstimateRouteFee](https://github.com/lightningnetwork/lnd/pull/8136) now
```
needs update

hieblmi commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,244 @@
+				lnrpc.PaymentFailureReason_FAILURE_REASON_INCORRECT_PAYMENT_DETAILS,
+			},
+		},
+	}
```
add edge case tests

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -356,20 +359,62 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+		return s.probeDestination(req.Dest, req.AmtSat)
+
+	case isProbeInvoice:
+		deadLine := time.Now().Add(s.cfg.FeeEstimationTimeout)
```
use user-provided timeout if provided

JssDWt commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -404,8 +449,241 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+						RoutingFeeMsat: paymentFee(
+							payment,
+						),
+						FailureReason: payment.FailureReason, //nolint:lll
```
Should this be `FAILURE_REASON_NONE`? And the other fields in the DestinationFeeEstimate should be set? `time_lock_delay`, `dest`, and `channel_id`.

JssDWt reviewed (commented):
Two small comments, otherwise looks good.

hieblmi commented:
There are some LSP heuristic improvements that are still pending. I will ping the reviewers once they are pushed.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+// hints don't indicate an LSP, they are passed as arguments to the
+// SendPayment_V2 method, which enable it to send probe payments to the payment
+// request destination.
+func (s *Server) probeInvoice(ctx context.Context, paymentRequest string,
```
Is `probePaymentRequest` a better name here?

ProofOfKeags replied:
yeah or probePaymentTarget?

ProofOfKeags commented on zpay32/hophint.go:
```diff
@@ -45,3 +45,13 @@ func (h HopHint) Copy() HopHint {
+// denoted in milli satoshi.
+func (h HopHint) HopFee(amountMSat int64) int64 {
+	feePpm := int64(h.FeeProportionalMillionths)
+	feeRateMsat := (amountMSat * feePpm) / 1_000_000
```
Add 1 here if the division remainder is nonzero. This estimation will underpay in 999_999 cases and depending on how strict the check is on the other side it may fall short.

hieblmi replied:
Nice catch!

ProofOfKeags replied:
Yeah @bitromortac is right. The [spec](https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#requirements-9) says that we don't really need to handle this rounding issue as nodes are recommended to accept payments that round down.

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	return resp, nil
+}
+
+// isLSP checks if the route hints indicate an LSP. An LSP is indicated in the
+// first return value if the last node in each route hint has the same node id.
+// If an LSP is indicated, the remaining hops in each route hint are returned
+// to the caller to be considered by the caller as route hints to the LSP node.
+// Additionally, the identified LSP node is returned as a separate hop hint.
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to obtain an upper bound of the fee estimate.
+func isLSP(routeHints [][]zpay32.HopHint) (bool, [][]zpay32.HopHint,
```
Ok on second thought it seems like the analysis context is a routing problem so I understand why this is the way it is.

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+				continue
+
+			case lnrpc.Payment_SUCCEEDED:
+				return nil, errors.New("probe payment " +
+					"unexpectedly succeeded")
+
```
I think this depends on hash preimages being broken or reused. Either way this would be really serious. Make the message scarier?

hieblmi replied:
😱

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -867,7 +1178,6 @@ func (s *Server) trackPayment(subscription routing.ControlTowerSubscriber,
 	identifier lntypes.Hash, stream Router_TrackPaymentV2Server,
 	noInflightUpdates bool) error {
 
-	// Stream updates to the client.
```
lol

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	hints := payReq.RouteHints
+	isLsp, lspAdjustedRouteHints, lspHopHint := isLSP(hints)
+	if !isLsp {
+		probeRequest.RouteHints = invoicesrpc.CreateRPCRouteHints(hints)
```
Why do we have to call the RPC here?

hieblmi replied:
Good question. Probably weird to call a private method with an rpc request param, but `sendProbePayment` uses it to directly forward it to the `SendPayment_V2`. We could introduce a struct instead to forward the rpc params, but then we'd have to convert it back to the rpc format within the `sendProbePayment` call to make it fit for `SendPayment_V2`.

ProofOfKeags replied:
Yeah I think you can get away with this, it just triggered my head scratching

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -356,20 +360,62 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+		return s.probeInvoice(probeCtx, req.PaymentRequest, req.Timeout)
+	}
+
+	return nil, errors.New("specify either a destination or an invoice")
```
I'd move this case to the top so we can make progressively more assumptions as the case statements go on.

hieblmi replied:
I think this fits into the default case of the switch here.

champo commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -356,20 +360,62 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
 
-// EstimateRouteFee allows callers to obtain a lower bound w.r.t how much it
-// may cost to send an HTLC to the target end destination.
+// EstimateRouteFee allows callers to obtain an upper bound w.r.t how much it
```
maybe I'm being a bit picky here, but how is the result guaranteed to be an upper bound? it's definitely not a lower bound (like query routes might be) but you can probably pay more if network conditions change between probe and attempt. imo it's an expected value.

hieblmi replied:
Fair, I was considering it an upper bound as we select the most expensive hop hint in case there are multiple route hints to choose from. But since the overall result depends on dynamic fee changes along a route I'll go with your suggestion.

champo commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -356,20 +360,62 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+// may cost to send an HTLC to the target end destination. This method sends
+// probe payments to the target node, based on target invoice parameters and a
+// random payment hash that makes it impossible for the target to settle the
+// htlc. If provided with a destination key and amount, this method will perform
+// a local graph based fee estimation.
```
this sounds like it should be documented clearly in the proto files. it would surprise me a lot! on a related note, as a caller I would expect to be able to choose whether I want probing or query routes when providing an invoice. for example for time-sensitive contexts where waiting 10s of seconds for a probe is not acceptable.

hieblmi replied:
Done.

champo commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+		AllowSelfPayment: false,
+		AmtMsat:          int64(*payReq.MilliSat),
+		PaymentHash:      paymentHash[:],
+		FeeLimitSat:      100_000_000,
```
nit: the query routes branch uses the same limit but written differently, a file-level constant or at least using the same constant would be nice

champo commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+
+	// The probe succeeded, so we can add the last hop's fee to the
+	// to fee the payment probe returned.
+	resp.RoutingFeeMsat += hopFee
```
Won't the returned `TimeLockDelay` be innacurate here? It should take into account the delay set in `lspHopHint`.

hieblmi replied:
nice catch !

champo commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+		// We are interested in an upper bound of a fee estimate when
+		// paying an LSP invoice. If the LSP hop hints indicate
+		// different final hop fees we track the highest.
+		if refHopHint.HopFee(1_000_000) < rhLastHop.HopFee(1_000_000) {
```
this could use the payment amount from the invoice to make it precise

hieblmi replied:
nice!

champo commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			"preimage: %w", err)
+	}
+
+	probeRequest := &SendPaymentRequest{
```
Shouldn't this also set `FinalCltvDelta` to match the invoice?

hieblmi replied:
done

champo commented on lnrpc/routerrpc/router_server_test.go:
```diff
@@ -214,3 +217,177 @@ func TestTrackPaymentsNoInflightUpdates(t *testing.T) {
+	require.True(t, reflect.DeepEqual(emptyRouteHints, adjustedRouteHints))
+	require.Nil(t, lspHint)
+
+	// A single hop should not be considered an LSP.
```
this comment does not match the test case below it

hieblmi replied:
"should be"

champo commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,226 @@
+			expectedRoutingFeesMsat: lspDiffExpectedMultiHintFeeMsat,
+			expectedFailureReasons:  lnrpc.PaymentFailureReason_FAILURE_REASON_NONE,
+		},
+	}
```
given that this work enables probing to LSPs, a test case where the private channel doesn't actually exist feels necessary. That would guarantee that it actually probes up to the LSP.

hieblmi replied:
@champo 
To clarify, the assumption is that LSPs block payment probes. So if our heuristic `isLsp` turns true we can't really tell if a private channel that is indicated in the hop hints actually exists or not. We will probe to the LSP regardless and manually determine the last hop's fee.
The test cases above already ensure that we are probing up to the Lsp(Bob) because the routes over Dave and Eve are high fee, so we expect the payment probe to go through Carol to Bob, which actually is the case since it matches our expected fees.

Let me know if you have further doubts or test suggestions here. Thanks for your review !

champo replied:
I'm getting a bit confused from your answer, so I'll recap what I understand from the harness and these test cases so it's easier to find out where the confusion is coming from:
1. There's an open private channel between Bob and Paula
2. Bob is the LSP
3. The tests expect the probe to be done up to Bob

Given all these if the code had a bug and decide to probe up to Paula, wouldn't this test case succeed too? Paula would be the one to return the incorrect payment details error which would be accepted as a sucessful probe.

hieblmi replied:
That makes sense, I got you. I added a test case with a node without any edges and use it as a destination of the probe, with a route hint through the previously used LSP. This should return the same results as the case where the private hop exists.

champo reviewed (commented):
Looks really good! It should be compatible with Muun's approach 😄

hieblmi commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,226 @@
+	//       \__ Dave ____/
+	//
+	req := &mppOpenChannelRequest{
+		amtAliceCarol: 235_000,
+		amtAliceDave:  135_000,
+		amtCarolBob:   135_000,
+		amtCarolEve:   135_000,
+		amtDaveBob:    135_000,
+		amtEveBob:     135_000,
+		amtBobPaula:   135_000,
```
Which aspect of the fee estimation should random channel sizes test for?

ProofOfKeags replied:
ah good point. It would make it clear that the outcome of the test does not depend on channel size, but perhaps the value of demonstrating that isn't worth the work to randomize them.

hieblmi commented on itest/lnd_mpp_test.go:
```diff
@@ -287,18 +297,43 @@ func (m *mppTestScenario) openChannels(r *mppOpenChannelRequest) {
+	// A function to check if a channel point is Paula's. This helps us to
+	// skip channel announcements for Paula's channel.
+	isPaulasChannel := func(cp string) bool {
+		resp := m.paula.RPC.ListChannels(&lnrpc.ListChannelsRequest{})
```
nice catch.

hieblmi commented on sample-lnd.conf:
```diff
@@ -1205,6 +1205,9 @@
 
+; Defines the the maximum duration that the probing fee estimation is allowed to
+; take.
+; routerrpc.fee-estimation-timeout=1m
```
I've chose the default timeout for payments from `lncli`, see https://github.com/lightningnetwork/lnd/blob/b76d3e64d32d1ea107942766c0194d4ec79ef7d6/cmd/lncli/cmd_payments.go#L33

hieblmi commented on docs/release-notes/release-notes-0.18.0.md:
```diff
@@ -175,9 +175,15 @@
   of confirmations) has been added to the `WalletBalance` RPC call.
+ 
+* [EstimateRouteFee](https://github.com/lightningnetwork/lnd/pull/8136) extends
+   the graph based estimation by a payment probe based approach which can lead 
```
Updated to sound a little less cumbersome.

bitromortac commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,229 @@
+
+	// Make sure Alice has heard it.
+	ht.AssertChannelPolicyUpdate(
+		mts.alice, mts.eve, expectedPolicy, chanPointCarolEve, false,
```
This should be `chanPointEveBob` to check the relevant channel?

bitromortac commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,229 @@
+	})
+	bobPaulaChanID := bobsPrivChannels.Channels[0].ChanId
+
+	// Increase Dave's fee to make the test deterministic. Otherwise, it
```
It would be clearer if you state that the paths over Eve and Dave are disabled due to their high fees. It would be cool if you could compare the same probe destination with the two different approaches, where the graph based method returns a different result than the probing based one. Say you would first try Carol-Bob, which doesn't work because of missing liquidity and then try Dave-Bob, which has higher fees but enough liquidity.

bitromortac commented on itest/lnd_mpp_test.go:
```diff
@@ -176,8 +178,8 @@ type mppTestScenario struct {
-//	Alice -- Carol ---- Bob
-//	    \              /
+//	Alice -- Carol ---- Bob ---------- Paula
+//	    \              /     (private)
```
it would be great if you could put these preparatory changes into its own commit

hieblmi replied:
done

bitromortac reviewed (commented):
Very nice additions @hieblmi :rocket:! I've got mostly nits and test improvement suggestions. Another question is how we should handle zero amount invoices.

champo commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			case lnrpc.Payment_FAILED:
+				// Incorrect payment details points to a
+				// successful probe.
+				if payment.FailureReason == lnrpc.PaymentFailureReason_FAILURE_REASON_INCORRECT_PAYMENT_DETAILS { //nolint:lll
```
should this check which node in the route returned the error?

hieblmi replied:
Will add a check to verify that the index of the hop that returned `incorrect payment details` is the last index of the route. In any other case we'll flag this as an probing error to the client.

ProofOfKeags commented on zpay32/hophint.go:
```diff
@@ -45,3 +49,28 @@ func (h HopHint) Copy() HopHint {
+
+	// Increase the fee rate by one milli satoshi if there was a remainder
+	// in the division.
+	if (amountMSat*feePpm)%1_000_000 > 0 {
+		feeRateMsat++
+	}
```
👍🏼

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	return resp, nil
+}
+
+// isLSP checks if the route hints indicate an LSP. An LSP is indicated in the
+// first return value if the last node in each route hint has the same node id.
+// If an LSP is indicated, the remaining hops in each route hint are returned
+// to the caller to be considered by the caller as route hints to the LSP node.
+// Additionally, the identified LSP node is returned as a separate hop hint.
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to obtain an upper bound of the fee estimate.
```
suggestion: i might recommend that we say somewhere here that it's a heuristic

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+						FailureReason:  lnrpc.PaymentFailureReason_FAILURE_REASON_NONE, //nolint:lll
+					}, nil
+				}
+
```
Yeah all switches should be exhaustive. If it should be "impossible" put a panic in there so we can be instantly aware of core assumption failures.

hieblmi replied:
For now I am returning the generic `FAILURE_REASON_ERROR` here, if no other return reason makes sense.

hieblmi replied:
Added all possible payment status cases plus a default case.

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+// Send sends a payment status update to a payment stream that the caller can
+// evaluate.
+func (p *probePaymentStream) Send(response *lnrpc.Payment) error {
+	p.stream <- response
```
I think we need a `select` here on `p.ctx.Done()` mixed in with this.

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	go func() {
+		err := s.SendPaymentV2(req, paymentStream)
+		if err != nil {
+			errChan <- err
```
moar selects

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to obtain an upper bound of the fee estimate.
+func isLSP(routeHints [][]zpay32.HopHint, amtMsat int64) (bool,
+	[][]zpay32.HopHint, *zpay32.HopHint) {
```
nonblocking: Following this double array of HopHints is a bit tough, I might suggest using types to package one of the layers into something like "Route" and implement some functions on that type to do the stripping you do in the body of this function.

hieblmi replied:
Initially I was looking for a `Route` struct, but once familiar with it, I found the implicit design nice. Adding a new type here but not in other places would also confuse, unless we changed it in all places.

hieblmi replied:
done

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -356,25 +364,72 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+		defer cancel()
+
+		return s.probePaymentRequest(
+			probeCtx, req.PaymentRequest, req.Timeout,
```
Agree, although I now I wonder if there is any value to the probe context in `sendProbePayment` as the payment itself is parametrized with the timeout.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			"preimage: %w", err)
+	}
+
+	amtMsat := int64(*payReq.MilliSat)
```
I'd think so, a probing amount `<= 0` doesn't make sense imo.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			refHopHint.FeeProportionalMillionths =
+				rhLastHop.FeeProportionalMillionths
+
+			refHopHint.CLTVExpiryDelta = rhLastHop.CLTVExpiryDelta
```
Are you suggesting that we modify the `refHopHint` to make it track the maximum `CLTVExpiryDelta`?

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			switch payment.Status {
+			case lnrpc.Payment_INITIATED:
+			case lnrpc.Payment_IN_FLIGHT:
+				continue
```
good point

hieblmi commented on lnrpc/routerrpc/router_server_test.go:
```diff
@@ -214,3 +217,151 @@ func TestTrackPaymentsNoInflightUpdates(t *testing.T) {
+func TestIsLsp(t *testing.T) {
+	probeAmtMsat := int64(1_000_000)
+
+	alicePrivKey, err := btcec.NewPrivateKey()
```
Is it really shorter than preparing for `func NewPublicKey(x, y *FieldVal) *PublicKey`?

hieblmi commented on lnrpc/routerrpc/router_server_test.go:
```diff
@@ -214,3 +217,151 @@ func TestTrackPaymentsNoInflightUpdates(t *testing.T) {
+		{},
+	}, probeAmtMsat)
+	require.False(t, isLsp)
+	require.True(t, reflect.DeepEqual(emptyRouteHints, adjustedRouteHints))
```
Added a table structure.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -356,25 +364,72 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+	isProbeDestination := len(req.Dest) > 0
+	isProbeInvoice := len(req.PaymentRequest) > 0
+
+	if !isProbeDestination && !isProbeInvoice {
```
this can be included in the switch and simplified to:
```
case isProbeDestination == isProbeInvoice:
		return nil, errors.New("specify either a destination or an " +
			"invoice")
```

hieblmi commented on zpay32/hophint.go:
```diff
@@ -2,6 +2,7 @@ package zpay32
 
 import (
 	"errors"
+
```
done

hieblmi commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,229 @@
+
+	// expectedRoutingFeesMsat are the expected routing fees that will be
+	// returned by the probe.
+	expectedRoutingFeesMsat int64
```
Thank you for the suggestions, I included them all except for the timeout case which I am still working on, but I think the review can continue without it.

hieblmi commented:
@bitromortac, @ProofOfKeags, @champo  thanks for your thorough reviews. I think I addressed all comments.

I am still working on the probing timeout test, but imo the review can continue without it for now.

@bitromortac 
We talked offline about the total cltv delta discrepancy of 3 blocks between graph and probe based estimates. Do harmonize the two I've added the block padding delta to the graph based estimation, see https://github.com/lightningnetwork/lnd/pull/8136/commits/87e0f9f218acd1d16ec0b45e75dc7d95be04fc58#diff-9ced6590b7c4cfad60e1fa1641b602da6fe758092211f30a55b19771e4186842R444

The other uncertainty I had was around handling the `MinFinalCLTVExpiryDelta` in combination with hop hints. In our LSP case we now add the cltv delta provided in the hop hint to the total cltv delta determined by the probe, but only in case it is greater than the `payReq.MinFinalCLTVExpiry()`, and `payReq.MinFinalCLTVExpiry()` otherwise, see https://github.com/lightningnetwork/lnd/pull/8136/commits/87e0f9f218acd1d16ec0b45e75dc7d95be04fc58#diff-9ced6590b7c4cfad60e1fa1641b602da6fe758092211f30a55b19771e4186842R550

I still have a doubt about the hop cltv determined by our `isLsp` heuristic, it is here https://github.com/lightningnetwork/lnd/pull/8136/commits/87e0f9f218acd1d16ec0b45e75dc7d95be04fc58#r1479710025

Thanks again for your time and effort!

hieblmi commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,323 @@
+package itest
```
TODO: Add test for probing in the the non-Lsp case.

bitromortac reviewed (commented):
Great tests/itests! I think the functionality added looks solid, the cltv calculation may need to be checked (see comment).

The issue with the zero amount invoices made me think if accepting an invoice to the api is really the best choice? There is some flow where a wallet gets a zero-amount invoice and the user fills in the amount. A probe would then be done with that amount. Right now we fail that use-case. Also there will be other invoice formats  (e.g. bolt 12) or use cases that we may need to support here. So the question is whether we should not just extend the proto with hop hints, final cltv, node features and a bool to indicate if we want to probe or not, leaving the api backwards compatible. The api user could decode the invoice before using this endpoint with https://lightning.engineering/api-docs/api/lnd/lightning/decode-pay-req. I think it wouldn't be a big change to this PR but could add more value (also the current graph-based approach would become better by being able to take hop hints into account)? The lncli command could stay the same with using the invoice for convenience. Would appreciate feedback for that @ all.

alexbosworth commented:
> The issue with the zero amount invoices made me think if accepting an invoice to the api is really the best choice? There is some flow where a wallet gets a zero-amount invoice and the user fills in the amount. A probe would then be done with that amount. Right now we fail that use-case. 

Failing isn't good, but it could also just have a separate amount field to mirror the send payment API, which would match the design pattern of probe with the arguments you want to use for a payment before you actually pay.

I can see the argument for generic probing fields but thinking about the history of this feature what motivated it from a user perspective is that doing this yourself is too hard, even though you can actually do everything yourself with the existing APIs. This API adds zero functionality you couldn't achieve some other way with other APIs and you can already break things down into components if you want to do that. So given that motivating logic, making it easier with just the payment request would be easier and therefore correct for this API.

saubyk commented:
> Failing isn't good, but it could also just have a separate amount field to mirror the send payment API, which would match the design pattern of probe with the arguments you want to use for a payment before you actually pay.
> 
> I can see the argument for generic probing fields but thinking about the history of this feature what motivated it from a user perspective is that doing this yourself is too hard, even though you can actually do everything yourself with the existing APIs. This API adds zero functionality you couldn't achieve some other way with other APIs and you can already break things down into components if you want to do that. So given that motivating logic, making it easier with just the payment request would be easier and therefore correct for this API.

+1
Also, if keeping the API future compatible with bolt12 is a concern, can that be handled internally by checking the format first for the invoice type? I believe a determination can be made my looking at the invoice string if it's bolt11 or 12

ProofOfKeags commented on lntest/node/harness_node.go:
```diff
@@ -876,6 +876,33 @@ func (hn *HarnessNode) RestoreDB() error {
+// expected policy for further check. It changes the node's `FeeBaseMsat` from
+// 1000 msat to 500,000 msat, and `FeeProportionalMillonths` from 1 msat to 1000
+// msat.
+func (hn *HarnessNode) UpdateGlobalPolicy(baseFeeMsat int64, feeRate float64,
```
This seems an odd choice as far as composability goes, shouldn't we pass in the `lnrpc.RoutingPolicy`? rather than making this function a constructor *and* a setter?

hieblmi replied:
yes totally, I didn't take a closer look when I was migrating this method. Now simplified it to a minimum I think.

ProofOfKeags commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -398,14 +441,350 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			routeHints[0][:len(routeHints[0])-1],
+		)
+	}
+	for i := 1; i < len(routeHints); i++ {
```
I'd like to see some of this broken up, into the separate checks. Right now all of the different heuristics are mixed together in the same loop which makes it hard to figure out what is being looked at. Instead, I'd recommend having a "node id homogeneity check", a "max fee selector", a "max cltv selector", etc. and then have the overall isLSP function be a list of these heuristics bundled. Especially with early exit, reasoning about for loops larger than a screen height is difficult, and when someone returns to these heuristics months from now they'll be grateful that things are neatly organized into their constituent pieces.

ProofOfKeags replied:
I don't think there's a ton of value to doing a golang interface for `heuristicCheck` but what I was saying is that we have three distinct things I was able to see. I'll provide signatures

```go
// note it is possible not all of these individual heuristics require all of these arguments, and
// the arguments should be trimmed if unused.

func HasHomogenousNodeIDs(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) bool { ... }

func FindMaxCLTV(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) uint32 { ... }

func FindMaxFee(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) lnwire.MilliSatoshi { ... }

func isLSP(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) *lspHeuristics {
    var heuristics lspHeuristics
    heuristics.sameNodeID = HasHomogenousNodeIDs(routeHints, amt)
    heuristics.maxCLTV = FindMaxCLTV(routeHints, amt)
    heuristics.maxFee = FindMaxFee(routeHints, amt)

    // maybe we do some further computation on these results together here in the future or something
}
```

This way we can look at `isLSP` and be like "Oh 🤔, that's what we are measuring to make our determinations", and each of those heuristics *only* measures one thing, making it easy to follow and verify it works correctly as designed. This way, if we want to add a new check, it's simply a matter of implementing a new standalone function for it, and then adding it to the bundle for `isLSP`, rather than trying to splice it into a large for-loop.

Hopefully this sheds light on the approach I'm going for? Lmk if you foresee issues here, but I think it will make the code substantially nicer to look at 😁

hieblmi replied:
Thanks that makes total sense. I am glad I asked as I was taking your "isLsp is a list of these heuristics" too literal by passing in a slice of these checks.

I'll go with your suggestion though I think the checks don't need to be exported.

ProofOfKeags commented on lnrpc/routerrpc/router_server_test.go:
```diff
@@ -214,3 +217,170 @@ func TestTrackPaymentsNoInflightUpdates(t *testing.T) {
+	bobExpensiveCopy.FeeBaseMSat = 1_000_000
+	bobExpensiveCopy.FeeProportionalMillionths = 1_000_000
+
+	lspTestCases := []struct {
```
🫡

ProofOfKeags reviewed (commented):
My biggest observation is that the isLSP heuristic function should be cleaned up from an organization perspective. I expect that over time that code will need to be serviced over and over again as the industry evolves, and having the different heuristics tightly organized is going to be key to ensure that the evolution can happen in a way that doesn't introduce subtle bugs as well as being rapidly iterated on as we come up with different ways to assess something that is fuzzy.

Otherwise I'm a big fan of the PR overall 😁

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -398,14 +441,350 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	for i := 1; i < len(routeHints); i++ {
+		rh := routeHints[i]
+
+		// Skip empty route hints.
+		if len(rh) == 0 {
+			continue
+		}
```
I don't see where this would have an impact. It indicates that the invoice creator messed something up for a single hint, as long as the other hints are functional we can go ahead and estimated imo.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -398,14 +441,350 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	}
+
+	lastAttempt := p.Htlcs[len(p.Htlcs)-1]
+	lastRoute := lastAttempt.Route
```
Added nil checks. As discussed offline there can be multiple attempts in `p.Htlcs`, however they contain failed payment attempts. Only the last one will be successful.

hieblmi commented on itest/lnd_amp_test.go:
```diff
@@ -55,8 +55,8 @@ func testSendPaymentAMPInvoiceCase(ht *lntest.HarnessTest,
 	//             /       \
-	// Alice -- Carol ---- Bob
-	//      \              /
+	// Alice -- Carol ---- Bob ------------ Paula
```
I've reverted the unneeded ones. We can create the additional nodes and private channels in the actual test.

hieblmi commented on itest/lnd_estimate_route_fee_test.go:
```diff
@@ -0,0 +1,365 @@
+				int64(routing.BlockPadding),
+			expectedFailureReasons: failureReasonNone,
+		},
+		// we expect the previous probing results adjusted by bob's hop
```
Addressed all the nits, the duplicate test comment was copy/paste remainder.

hieblmi commented on sample-lnd.conf:
```diff
@@ -1205,6 +1205,9 @@
 ; `Payment_In_FLIGHT` will be sent for compatibility concerns.
 ; routerrpc.usestatusinitiated=false
 
+; Defines the the maximum duration that the probing fee estimation is allowed to
```
done

hieblmi commented:
Once again thanks for the thorough review @ProofOfKeags + @bitromortac, I've addressed all your comments.
Most notably I restructured `isLsp` to only check what its name implies, and added `prepareLspRouteHints` to handle the extraction of `adjustedRouteHints` which allow probing the LSP. I've also made the requested changes to tests and itests. 

This PR is ready for another review round.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -398,14 +445,399 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+func maxLspFee(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) (uint32,
+	uint32) {
+
+	var maxFeePpm uint32
```
nit: put in `var` block

bitromortac reviewed (commented):
I have some more nits and the itests needs an update to test real probing to the receiver which needs to take the bits of the destination node into account (read from the invoice as discussed).

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -398,14 +445,399 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+
+	// Create the LSP hop hint. We are probing for the worst case fee and
+	// cltv delta. So we look for the max values amongst all LSP hop hints.
+	refHint := routeHints[0][len(routeHints[0])-1]
```
I don't think we need to return an error but `false`? If the hints are `nil` they will be safely processed by `sendPayment`. That's tested in an itest `probe based estimate, empty route hint`.

bitromortac replied:
right, `isLSP` would handle this, but the question is where do you draw the line of checking for nil slices if you access those (or other situations where panics could happen). for me it would be within the scope of a function, as it's a reusable piece of code.

hieblmi replied:
apologies, I thought the mentioned line is in `isLsp` but it is in `prepareLspRouteHints`. Agree with what your are saying, will change.

bitromortac commented on lnrpc/routerrpc/router.proto:
```diff
@@ -353,14 +353,37 @@ message TrackPaymentsRequest {
+
+    /*
+    A user preference of how long a probe payment should maximally be allowed to
+    take, denoted in seconds. Although the probing function is aborted if the
+    timeout is reached, the underlying payment attempt is not cancelled. This
+    might leave the node with pending outgoing htlcs that wait for resolution.
```
nit: we should be more accurate here, suggestion:
`The probing payment loop is aborted if this timeout is reached. Note that the probing process itself can take longer than the timeout if the HTLC becomes delayed or stuck. Canceling the context of this call will not cancel the payment loop, the duration is only controlled by the timeout parameter.`

bitromortac reviewed (approved):
LGTM :tada:. There's only one remark left concerning the api docs.

hieblmi commented:
Once again thank you @bitromortac for this thorough review journey. The handling of route hints by `lnd` has been especially insightful for me.
